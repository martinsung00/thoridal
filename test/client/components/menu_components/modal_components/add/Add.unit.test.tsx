import React from "react";
import renderer from "react-test-renderer";
import ModalContent from "../../../../../../src/client/components/menu_components/modal_components/add/Add";
import { shallow, configure, ShallowWrapper } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Axios from "axios";
import MockAdapater from "axios-mock-adapter";

configure({ adapter: new Adapter() });

describe("Menu Option Add: Modal Content Tests", function () {
  // Fake Props for testing purposes
  const props = {
    toggleShow: jest.fn(),
    show: true,
  };

  afterEach(function () {
    props.show = true;
  });

  it("should render all appropriate elements as is", function () {
    const container:
      | renderer.ReactTestRendererJSON
      | renderer.ReactTestRendererJSON[]
      | null = renderer.create(<ModalContent {...props} />).toJSON();

    expect(container).toMatchSnapshot();
  });

  it("should change pages according to the current page state", function () {
    const wrapper: ShallowWrapper<
      any,
      Readonly<{}>,
      React.Component<{}, {}, any>
    > = shallow(<ModalContent {...props} />);

    wrapper.setState({
      current_page: "checkout",
      errored: false,
      trade_id: "",
    });

    expect(wrapper).toMatchSnapshot();

    wrapper.setState({
      current_page: "confirm",
    });

    expect(wrapper).toMatchSnapshot();

    wrapper.setState({
      current_page: "failed",
    });

    expect(wrapper).toMatchSnapshot();

    wrapper.setState({
      current_page: "undo-confirm",
    });

    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

  describe("Confirm Page", function () {
    let wrapper: ShallowWrapper<
      any,
      Readonly<{}>,
      React.Component<{}, {}, any>
    >;

    beforeEach(function () {
      wrapper = shallow(<ModalContent {...props} />);
    });

    afterEach(function () {
      wrapper.unmount();
    });

    it("should undo the recent add action", function (preventDefault) {
      wrapper.setState({
        current_page: "confirm",
        errored: false,
        trade_id: "abc",
      });

      wrapper.find("#undo-button").simulate("click", { preventDefault });

      expect(wrapper).toMatchSnapshot();
    });

    it("should close the modal", function (preventDefault) {
      wrapper.setState({
        current_page: "confirm",
        errored: false,
        trade_id: "abc",
      });

      wrapper.find("#close-button").simulate("click", { preventDefault });

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("Undo Confirmation Page", function () {
    it("should close the modal", function (preventDefault) {
      const wrapper: ShallowWrapper<
        any,
        Readonly<{}>,
        React.Component<{}, {}, any>
      > = shallow<ModalContent>(<ModalContent {...props} />);

      wrapper.setState({
        current_page: "undo-confirm",
        errored: false,
        trade_id: "abc",
      });

      wrapper.find("#close-button").simulate("click", { preventDefault });

      expect(wrapper).toMatchSnapshot();

      wrapper.unmount();
    });
  });

  describe("Failed Page", function () {
    it("should return to the previous state", function (done) {
      const wrapper: ShallowWrapper<
        any,
        Readonly<{}>,
        React.Component<{}, {}, any>
      > = shallow<ModalContent>(<ModalContent {...props} />);

      wrapper.setState({
        current_page: "failed",
        errored: true,
        trade_id: "",
      });

      expect(wrapper).toMatchSnapshot();

      wrapper.unmount();

      done();
    });
  });

  describe("Next State Method", function () {
    it("should change the current state", function (done) {
      const wrapper = shallow<ModalContent>(<ModalContent {...props} />);

      wrapper.setState({
        current_page: "checkout",
        errored: false,
        trade_id: "",
      });

      wrapper.instance()._nextState();

      expect(wrapper.state()).toEqual({
        current_page: "confirm",
        errored: false,
        trade_id: "",
      });

      wrapper.unmount();

      done();
    });
  });

  describe("Previous State Method", function () {
    const event = {
      currentTarget: {},
    } as React.MouseEvent<HTMLInputElement>;

    beforeAll(function () {
      event.preventDefault = jest.fn();
    });

    afterAll(function () {
      jest.resetAllMocks();
    });

    it("should change the current state", function (done) {
      const wrapper = shallow<ModalContent>(<ModalContent {...props} />);

      wrapper.setState({
        current_page: "confirm",
        errored: false,
        trade_id: "",
      });

      wrapper.instance()._prevState(event);

      expect(wrapper.state()).toEqual({
        current_page: "checkout",
        errored: false,
        trade_id: "",
      });

      wrapper.unmount();

      done();
    });
  });

  describe("Handle Eror Method", function () {
    it("should change the current state", function (done) {
      const wrapper = shallow<ModalContent>(<ModalContent {...props} />);

      wrapper.instance()._handleError();

      expect(wrapper).toMatchSnapshot();

      wrapper.unmount();

      done();
    });
  });

  describe("Handle Eror Method", function () {
    it("should change the current state", function (done) {
      const wrapper = shallow<ModalContent>(<ModalContent {...props} />);

      wrapper.instance()._handleError();

      expect(wrapper).toMatchSnapshot();

      wrapper.unmount();

      done();
    });
  });

  describe("Confirm ID Method", function () {
    it("should change the current state", function (done) {
      const wrapper = shallow<ModalContent>(<ModalContent {...props} />);
      const testID = "test";

      wrapper.setState({
        current_page: "checkout",
        errored: false,
        trade_id: "",
      });

      wrapper.instance()._confirmID(testID);

      expect(wrapper.state()).toEqual({
        current_page: "checkout",
        errored: false,
        trade_id: "test",
      });

      wrapper.unmount();

      done();
    });
  });

  describe("Confirm Undo Method", function () {
    it("should change the current state to undo-confirm", function (done) {
      const wrapper = shallow<ModalContent>(<ModalContent {...props} />);

      wrapper.setState({
        current_page: "confirm",
        errored: false,
        trade_id: "",
      });

      wrapper.instance()._handleUndoConfirm();

      expect(wrapper.state()).toEqual({
        current_page: "undo-confirm",
        errored: false,
        trade_id: "",
      });

      wrapper.unmount();

      done();
    });
  });

  describe("Handle Undo Method", function () {
    const flushPromises = () => new Promise(setImmediate);
    const event = {
      currentTarget: {},
    } as React.FormEvent<HTMLInputElement>;

    beforeAll(function () {
      event.preventDefault = jest.fn();
    });

    afterAll(function () {
      jest.resetAllMocks();
    });

    it("should change state to failed if the trade ID is missing", function (done) {
      const wrapper = shallow<ModalContent>(<ModalContent {...props} />);

      wrapper.setState({
        current_page: "checkout",
        errored: false,
        trade_id: "",
      });

      wrapper.instance()._handleUndo(event);

      expect(wrapper.state()).toEqual({
        current_page: "failed",
        errored: true,
        trade_id: "",
      });

      wrapper.unmount();

      done();
    });

    it("should change state failed if server encounters an error", async function (done) {
      const mock = new MockAdapater(Axios);
      const wrapper = shallow<ModalContent>(<ModalContent {...props} />);
      const spy = jest.spyOn<any, any>(wrapper.instance(), "_handleError");

      wrapper.setState({
        current_page: "confirm",
        errored: false,
        trade_id: "TEST",
      });

      mock.onDelete("/trade/id/TEST/delete").reply(500);

      try {
        wrapper.instance()._handleUndo(event);
        await flushPromises();
        wrapper.update();
      } catch (err) {
        expect(wrapper).toMatchSnapshot();
        expect(spy).toHaveBeenCalled();
      }

      wrapper.unmount();
      mock.restore();

      done();
    });

    it("should change state to undo confirmation if delete is successful", async function (done) {
      const mock = new MockAdapater(Axios);
      const wrapper = shallow<ModalContent>(<ModalContent {...props} />);

      wrapper.setState({
        current_page: "confirm",
        errored: false,
        trade_id: "TEST",
      });

      mock
        .onDelete("/trade/id/TEST/delete")
        .reply(200, { data: { rows: [{ id: "TEST" }] } });

      wrapper.instance()._handleUndo(event);

      expect(wrapper).toMatchSnapshot();

      wrapper.unmount();
      mock.restore();

      done();
    });
  });
});
