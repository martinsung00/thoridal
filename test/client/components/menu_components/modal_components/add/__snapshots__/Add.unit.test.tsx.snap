// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Option Add: Modal Content Tests Confirm Page should close the modal 1`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests Confirm Page should undo the recent add action 1`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests Failed Page should return to the previous state 1`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests Handle Eror Method should change the current state 1`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests Handle Eror Method should change the current state 2`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests Handle Undo Method should change state to undo confirmation if delete is successful 1`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests Undo Confirmation Page should close the modal 1`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests should change pages according to the current page state 1`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests should change pages according to the current page state 2`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests should change pages according to the current page state 3`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests should change pages according to the current page state 4`] = `ShallowWrapper {}`;

exports[`Menu Option Add: Modal Content Tests should render all appropriate elements as is 1`] = `
<div>
  <div>
    <form
      className="inputForm"
      onSubmit={[Function]}
    >
      <label>
        ID:
        <input
          id="id"
          name="id"
          onChange={[Function]}
          placeholder="e.g 1234ABC"
          required={true}
          type="string"
          value=""
        />
      </label>
      <div />
      <label>
        Company Name:
        <input
          id="company_name"
          name="company_name"
          onChange={[Function]}
          placeholder="e.g abc"
          required={true}
          type="string"
          value=""
        />
      </label>
      <div />
      <label>
        Ticker:
        <input
          id="ticker"
          name="ticker"
          onChange={[Function]}
          placeholder="e.g abc"
          required={true}
          type="string"
          value=""
        />
      </label>
      <div />
      <label>
        Reference Number:
        <input
          id="reference_number"
          name="reference_number"
          onChange={[Function]}
          placeholder="e.g 1234ABC"
          required={true}
          type="string"
          value=""
        />
      </label>
      <div />
      <label>
        Unit Price:
        <input
          id="unit_price"
          name="unit_price"
          onChange={[Function]}
          placeholder="e.g 19.99"
          required={true}
          type="number"
          value={0}
        />
      </label>
      <div />
      <label>
        Quantity:
        <input
          id="quantity"
          name="quantity"
          onChange={[Function]}
          placeholder="e.g 100"
          required={true}
          type="number"
          value={0}
        />
      </label>
      <div />
      <label>
        Total Cost:
        <input
          id="total_cost"
          name="total_cost"
          onChange={[Function]}
          placeholder="e.g 199.99"
          required={true}
          type="number"
          value={0}
        />
      </label>
      <div />
      <label>
        Trade Type:
        <select
          id="trade_type"
          onChange={[Function]}
          value="short"
        >
          <option
            value="long"
          >
            Long
          </option>
          <option
            value="short"
          >
            Short
          </option>
        </select>
      </label>
      <div />
      <label>
        Trade Status:
        <select
          id="trade_status"
          onChange={[Function]}
          value="not-fulfilled"
        >
          <option
            value="fulfilled"
          >
            Fulfilled
          </option>
          <option
            value="not-fulfilled"
          >
            Not Fulfilled
          </option>
        </select>
      </label>
      <div />
      <label>
        Trade Action:
        <select
          id="trade_action"
          onChange={[Function]}
          value="bought"
        >
          <option
            value="bought"
          >
            Bought
          </option>
          <option
            value="sold"
          >
            Sold
          </option>
        </select>
      </label>
      <div />
      <label>
        Note:
        <input
          id="note"
          name="note"
          onChange={[Function]}
          placeholder="e.g Hello World"
          required={true}
          type="string"
          value=""
        />
      </label>
      <div />
      <input
        id="submit-button"
        type="submit"
        value="Submit"
      />
    </form>
    <div
      id="cancel-button"
      onClick={[Function]}
    >
      <div>
        Cancel
      </div>
    </div>
  </div>
</div>
`;
